pylearn2/datasets/hdf5.py

```diff
@@ -2,338 +2,472 @@
-        if self.y is None:
-            space = X_space
-            source = X_source
+        data = alias_dict()
+        if use_h5py:
+            for s, a in safe_zip(sources, aliases):
+                if load_all:
+                    data[s, a] = self._fhandler[s][:]
+                else:
+                    data[s, a] = self._fhandler[s]
+                    # hdf5 handle has no ndim
+                    data[s].ndim = len(data[s].shape)
         else:
-            if self.y.ndim == 1:
-                dim = 1
-            else:
-                dim = self.y.shape[-1]
-            y_space = VectorSpace(dim=dim)
-            y_source = 'targets'
-            space = CompositeSpace((X_space, y_space))
-            source = (X_source, y_source)
-
-        self.data_specs = (space, source)
-        self.X_space = X_space
-        self._iter_data_specs = (X_space, X_source)
+            for s, a in safe_zip(sources, aliases):
+                if load_all:
+                    data[s, a](self._fhandler.getNode('/', s)[:])
+                else:
+                    data[s, a] = self._fhandler.getNode('/', s)
+        return data
 
+    @wraps(Dataset.iterator, assigned=(), updated=())
+    def iterator(self, mode=None, batch_size=None, num_batches=None,
+                 rng=None, data_specs=None, return_tuple=False,
+                 **kwargs):
+        """
+        data_specs : tuple
+            A `(space, source)` tuple. See :ref:`data_specs` for a full
+            description.
+        return_tuple : bool, optional
+            Always return a tuple, even if there is exactly one source
+            of data being returned. Defaults to `False`.
+        """
+        [mode, batch_size, num_batches, rng, data_specs] = self._init_iterator(
+            mode, batch_size, num_batches, rng, data_specs)
+        convert = None
+
+        if data_specs is None:
+            sources = self._get_canonical_sources()
+            spaces = self._get_canonical_spaces()
+            data_specs = (spaces, sources)
+            data = self._get_canonical_data()
+        else:
+            sources = data_specs[1]
+            data = [self.data[s] for s in sources]
+
+        return FiniteDatasetIterator(self,
+                                     mode(data[0].shape[0],
+                                          batch_size,
+                                          num_batches,
+                                          rng),
+                                     data_specs=data_specs,
+                                     return_tuple=return_tuple,
+                                     convert=convert)
+
+    def _get_canonical_sources(self):
+        """
+        Returns a list. The first element will be the source of `X` if present,
+        or of `topo_view` otherwise. The second element will be the source of
+        `y` if present. This is needed for compatibility with legacy code but
+        you should not rely on it.
+        """
```

Comment:
Oh, OK. I really don't think this is necessary to have this.
What is the problem if we don't, and force the user to always specify `data_specs` in `iterator()`?